src/aerospike_clj/client.clj

```diff
@@ -121,29 +124,64 @@
     (throw (Exception. (format "key is too long: %s..." (subs k 0 40)))))
   (Key. aero-namespace set-name k))
 
+(defn- ^Bin create-bin [^String bin-name bin-value]
+  (when (< MAX_BIN_NAME_LENGTH (.length bin-name))
+    (throw (Exception. (format "%s is %s characters. Bin names have to be <= 14 characters..." bin-name (.length bin-name)))))
+  (Bin. bin-name bin-value))
+
+(defn- ^Bin set-bin-as-null [^String bin-name]
+  (when (< MAX_BIN_NAME_LENGTH (.length bin-name))
+    (throw (Exception. (format "%s is %s characters. Bin names have to be <= 14 characters..." bin-name (.length bin-name)))))
+  (Bin/asNull bin-name))
+
 ;; get
 (defrecord AerospikeRecord [payload ^Integer gen ^Integer ttl])
 
 (defn- record->map [^Record record]
   (and record
-       (->AerospikeRecord
-         (get (.bins ^Record record) "")
-         ^Integer (.generation ^Record record)
-         ^Integer (.expiration ^Record record))))
+    (let [bins      (into {} (.bins ^Record record)) ;; converting from java.util.HashMap to a Clojure map
+          bin-names (mapv first bins)]
+      (->AerospikeRecord
+        (if (utils/single-bin? bin-names)
+          (utils/desanitize-bin-value (get bins ""))
+          (reduce-kv (fn [m k v]
+                       (assoc m (keyword k) (utils/desanitize-bin-value v)))
+            {}
+            bins))
+        ^Integer (.generation ^Record record)
+        ^Integer (.expiration ^Record record)))))
+
+(defn- map->multiple-bins [^IPersistentMap m]
+  (let [bins (for [[k v] m]
+               (create-bin (name k) (utils/sanitize-bin-value v)))]
+    (into-array Bin bins)))
+
+(defn- data->bins
+  "Function to identify whether `data` will be stored as a single or multiple bin record.
+  Only Clojure maps will default to multiple bins. Nested data structures are supported."
+  [data]
+  (if (map? data)
+    (map->multiple-bins data)
+    (into-array Bin [^Bin (Bin. "" (utils/sanitize-bin-value data))])))
 
 (defn get-single
   "Returns a single record: `(transcoder AerospikeRecord)`. The default transcoder is `identity`.
   Pass a `:policy` in `conf` to use a non-default `ReadPolicy`"
-  ([db index set-name] (get-single db index set-name {}))
-  ([db index set-name conf]
+  ([db index set-name] (get-single db index set-name [:all] {}))
+  ([db index set-name ^IPersistentVector bin-keys] (get-single db index set-name bin-keys {}))
+  ([db index set-name ^IPersistentVector bin-keys conf]
```

Comment:
I think initially, I did want to do this as well and then chose to default to Clojure keywords. I think there are two options here:
- ensure users only use string keys in this library (error when they don't). Probably better since that is how Aerospike wants it
OR
- convert every key into a string and expect them to call it back as such. Pretty much `str` everything in and out to be safe

Reply:
I'll go for the error option. Less surprising, more perfomant